Refine multiselect.

jasmussen opened the pull request:
This restores the native selection in addition to the borders.

Based on feedback by @ellatrix in https://github.com/WordPress/gutenberg/pull/19094#issuecomment-565401539 here's some refinement. GIFS:

![twenty](https://user-images.githubusercontent.com/1204802/70796566-e7422280-1da2-11ea-9cb3-1187391f0aea.gif)

![dark](https://user-images.githubusercontent.com/1204802/70796568-e8734f80-1da2-11ea-8451-13b50514e5da.gif)

![purple](https://user-images.githubusercontent.com/1204802/70796570-e9a47c80-1da2-11ea-9d9d-0a5f21094299.gif)

> Why not leave the text selection for extra clarity (it's really visible, works on all kinds of backgrounds)?

This is both a solid point, but also something that was the intent of the #19094. We can't always rely on the text selection marker being sufficient to indicate selection, for example a Cover block with no text will not show as selected because it has only a background. 19094 took the extreme and removed the text selection entirely, due to the idea that having _two_ selection indicators might be confusing. 

The efforts are inspired by these mockups:

<img width="1440" alt="69957382-de2f9680-1502-11ea-890b-9ed4c9a07ffe" src="https://user-images.githubusercontent.com/1204802/70796795-73ece080-1da3-11ea-9e9a-2689598388af.png">

<img width="1440" alt="69957443-f7384780-1502-11ea-84bb-0865414f7ffa" src="https://user-images.githubusercontent.com/1204802/70796796-751e0d80-1da3-11ea-9dd8-9d45b9c3f630.png">

In those, an exception is made for consecutive paragraphs, which if combined with improvements to the writing flow would be just delightful. But as soon as selection enters a non-text block, it converts to a fully bordered selection instead. The reason being that when you _copy_, you copy the blocks, not just the text inside.

So going back to the question, how do we ensure the selection color is visible on all backgrounds? We can't use `mix-blend-mode` or `filter`, as there are no colors that always are contrasty to their backgrounds, regardless of configuration. 

Even the native selection color is not always visible. Here, the editor background is the same color as the MacOS selection color:

![Screenshot 2019-12-13 at 12 30 00](https://user-images.githubusercontent.com/1204802/70797211-55d3b000-1da4-11ea-8c7d-de9b38d9fc7d.png)

You can _barely_ make out the selection. 

What does Figma do?

By default they have a lovely blue, like us:

![Screenshot 2019-12-13 at 12 28 05](https://user-images.githubusercontent.com/1204802/70797242-6552f900-1da4-11ea-82aa-2f6195bcd4d0.png)

What if the background is that blue?

![Screenshot 2019-12-13 at 12 28 20](https://user-images.githubusercontent.com/1204802/70797250-6ab04380-1da4-11ea-80aa-13cef3ab65be.png)

Only the rectangles are visible. So maybe two colors is an option, I'll follow up with a comment.

jasmussen committed:
Refine multiselect.

This restores the native selection in addition to the borders.

Additionally, it changes the cross-block selection border to also have a white outline.

jasmussen commented:
Here's what happens if we add a 1px white outline outside the blue one:

![Screenshot 2019-12-13 at 12 34 34](https://user-images.githubusercontent.com/1204802/70797690-4f920380-1da5-11ea-9df0-e5821455d81b.png)

![Screenshot 2019-12-13 at 12 34 59](https://user-images.githubusercontent.com/1204802/70797699-50c33080-1da5-11ea-9e21-3a1130daaf33.png)

![Screenshot 2019-12-13 at 12 35 35](https://user-images.githubusercontent.com/1204802/70797708-528cf400-1da5-11ea-90b1-10f4a9d2b48f.png)

The latter shows a theme that accidentally has the same color as the selection outline. ¯\_(ツ)_/¯

What do you think?

ellatrix commented:
@jasmussen Thanks for exploring all the options.

I don't really _mind_ hiding the native selection. The most important part for me is to use native selection underneath to power block multi selection because it's really smooth. I don't really care so much how it's styled or if it's hidden, but I do wonder what is most clear and visible. It appears that both have pros and con, and I agree that having both is a bit strange.

Let me think about this for a bit. :D

jasmussen commented:
> I don't really mind hiding the native selection. The most important part for me is to use native selection underneath to power block multi selection because it's really smooth

I agree, this is still present, it's only a visual change. 

I pushed and rebased and added a white extra border. Here's how it looks:

![yellow](https://user-images.githubusercontent.com/1204802/70799073-f035f280-1da8-11ea-941d-3ba939463ccb.gif)

![white](https://user-images.githubusercontent.com/1204802/70799076-f1671f80-1da8-11ea-86f6-2d41204820b4.gif)

![dark](https://user-images.githubusercontent.com/1204802/70799080-f2984c80-1da8-11ea-9cfe-e86cb1231259.gif)

ellatrix commented:
I like that. With the extra white border (love it), how do you feel about the text selection? Should it stay or should it go? I don't have an opinion.

ellatrix commented on packages/block-editor/src/components/block-list/style.scss:
```diff
@@ -182,7 +182,7 @@
 		// Show selection borders around every non-nested block's actual footprint.
 		> .block-editor-block-list__block-edit > [data-block] {
-			box-shadow: 0 0 0 2px #057cba;
+			box-shadow: 0 0 0 2px $blue-blueberry, 0 0 0 4px $white;
```
Love this. Guaranteed to see something.

ellatrix replied:
I played a bit with a dashed line (white and blue), but it's a bit too busy. Perhaps if the dashed lines could be longer, it would work better (maybe border image). In any case this is looking good.

ellatrix reviewed (approved):
Awesome. I like the blueberry colour too. :)

jasmussen commented:
> I like that. With the extra white border (love it), how do you feel about the text selection? Should it stay or should it go? I don't have an opinion.

If, one day, we can make it so you can select partial text between two subsequent paragraphs and delete to merge, then the text selection marker should be visible ony between those two, like [in this image](https://user-images.githubusercontent.com/1204802/70796795-73ece080-1da3-11ea-9e9a-2689598388af.png).

If we were to merge this in as a 7.1.1 improvement, which maybe isn't necessary, then I think it should stay across all blocks as it is.

Otherwise I think the text selection marker should remain only inside text.

ellatrix commented:
@jasmussen Feel free to remove or leave the text selection.

ellatrix commented:
(The border change is a great improvement on its own.)

jasmussen commented:
Thanks for the review. I will wait for a few more thoughts on the text selection, then merge after!

jasmussen committed:
Refine colors, add dark theme exception.

jasmussen commented:
Alright, I've had a good long think about this, and I've changed my mind on a few things. 

- The Blueberry color is my favorite, but it's too early to introduce it like this in isoluation. Let's revisit this one.
- The white border is effective, but it's also a confusing double border and heavy UI. I've changed so that the color is dependant on whether the theme is light (default) or dark (opt-in), and the selection color is a crisp white when in dark theme. 

I think it looks and balanced.

![twenty](https://user-images.githubusercontent.com/1204802/70812107-444fcf80-1dc7-11ea-940c-50910c11e513.gif)

![white](https://user-images.githubusercontent.com/1204802/70812111-4580fc80-1dc7-11ea-905f-5e4b83495e9d.gif)

![dark](https://user-images.githubusercontent.com/1204802/70812115-47e35680-1dc7-11ea-99ed-184cbbf6c763.gif)

@ellatrix feel free to re-review. I realize it's a big change.

ellatrix commented:
I trust you with this stuff! Feel free to merge. :) I do really like the blueberry but yeah maybe not the right time for it. The theme setting sounds good too.

jasmussen commented:
>  I do really like the blueberry but yeah maybe not the right time for it

Thanks. I'm not over the blueberry yet, I hope to see it return in style! Appreciate the re-review. Have a great weekend!

jasmussen commented:
Your feedback is very valid, @aduth!

What so excited me about Ella's work was that it as the first substantial step towards breaking the barrier between paragraphs, potentially (but still a technical hurdle) so that you could select from the middle of one paragraph, into the next, press Delete, and merge the two. We'd basically make an exception for multi-select so that so long as you select in paragraphs, it's one continuous text selection, like [this](https://user-images.githubusercontent.com/1204802/70796795-73ece080-1da3-11ea-9e9a-2689598388af.png). Only when you then select into a non-text block would it convert into [this](https://user-images.githubusercontent.com/1204802/70796796-751e0d80-1da3-11ea-9dd8-9d45b9c3f630.png).

So why bring in the border around each block? A couple of reasons:

1. The dark left border has shown its limitations in practice, exemplified by selecting child blocks in things like the Navigation or Social Links blocks. It is easily mistaken for _focus_, which it isn't.
2. The left border is not visible for full-wide blocks.
3. The majority of blocks selected using the text marker are not visibly selected, because things like backgrounds and non-text elements are not necessarily highlighted. 

Example of 3:

![Screenshot 2019-12-13 at 18 33 56](https://user-images.githubusercontent.com/1204802/70819740-428e0800-1dd7-11ea-81e0-28376a3a977e.png)

This is a little misleading as it might suggest the thing you copy, if you copy/paste, is "Text Cover Text", when in fact you're copying the entire markup of a sequence of three blocks. 

Finally, if we _do_ manage to break the boundary between consecutive Paragraph blocks, we'll still want to have movers available, and unless we actually show some sort of boundary, you might think that only the partial selection is what gets moved, forcing a split or something. Theoretically we'd want to do something like this to indicate what actually gets moved:

<img width="1440" alt="Multi Selection, Between Text Blocks" src="https://user-images.githubusercontent.com/1204802/70819947-bf20e680-1dd7-11ea-929f-462e00f8aee2.png">

I understand what excitement the native selection brought, I share it. I still think magical things could be in our future, but I also think we have to take mindful steps towards it. And in the interim, to solve the problems outlined above, I believe the borders are necessary.

aduth commented:
I think the biggest change for me is that previously, it felt as though there was a continuation between the two paragraphs which which helped to reinforce a seamless feel associated with some future of partial paragraph selections. In other words, if the goal is partial paragraph selections, v7.1.0 felt like a large step forward, but the more recent changes feel like small steps backward again (at least for the interaction between paragraphs).

I think the last image you shared in your comment would be a reasonable compromise to retain this sense of continuity. If that's what you're suggesting be the direction, I can get on board with that.

jasmussen commented:
> I think the last image you shared in your comment would be a reasonable compromise to retain this sense of continuity. If that's what you're suggesting be the direction, I can get on board with that.

The last mockup is that I'm _hoping_/dreaming of us accomplishing. Absolutely.

mtias commented:
> Just one person's feedback: I quite like how the multi-selection works across paragraph blocks in v7.1.0 of the plugin, where it feels much more like working with a seamless sequence of text moreso than a collection of distinct blocks.

The biggest issue for me was from a usability and expectations perspective — it seemed like you had selected text across blocks, but in reality the blocks themselves as a whole were selected. Actions like backspace or moving up / down would apply to the entire blocks, which wasn't very clearly communicated.

ellatrix commented:
We can have both things. No border in the future for blocks that are partially selectable, and border for the rest. I’ll work on partial selection sometime soon.

mtias commented:
Yes, once we have the mechanisms in place we can tweak the presentation to make it always clear what is happening.
